tmt/base.py

```diff
@@ -382,29 +440,56 @@ def _save_context(cls, context):
                 pattern = os.path.join('/', os.path.relpath(current, root))
                 # Prevent matching common prefix from other directories
                 pattern = f"{pattern}(/|$)"
+            assert cls._context is not None  # narrow type
             cls._context.params['names'] = tuple([
                 pattern if name == '.' else name for name in names])
 
-    def name_and_summary(self):
+    def name_and_summary(self) -> str:
         """ Node name and optional summary """
         if self.summary:
             return '{0} ({1})'.format(self.name, self.summary)
         else:
             return self.name
 
-    def ls(self, summary=False):
+    def ls(self, summary: bool = False) -> None:
         """ List node """
         echo(style(self.name, fg='red'))
         if summary and self.summary:
             echo(tmt.utils.format('summary', self.summary))
 
-    def export(self, format_='yaml', keys=None):
+    @overload
+    def export(self, *, format_: Literal[ExportFormat.DICT],
+               keys: Optional[List[str]] = None) -> Dict[str, Any]:
+        pass
+
+    @overload
+    def export(self, *, format_: Literal[ExportFormat.YAML],
+               keys: Optional[List[str]] = None) -> str:
+        pass
+
+    @overload
+    def export(self, *, format_: Literal[ExportFormat.RST],
+               keys: Optional[List[str]] = None) -> str:
+        pass
+
+    @overload
+    def export(self, *, format_: Literal[ExportFormat.NITRATE],
```

Comment:
Because the `format_` parameter is of type `ExportFormat`, and that does include even formats the actual implementation does not support. There's no way to express this with just `ExportFormat`, we would need yet another type with just supported formats, and that seems like overkill.

In general, I consider the enum to be a temporary compromise. Since the export functionality might target different systems, I'd envision it as a more pluggable system, e.g. with plugins registering their "format" & a handler to perform the export. That would mitigate the need for output annotations depending on the output, and with a bit of abstraction, the enum & `@overload`s as well.

Reply:
I see, thanks much for the detailed clarification!